rfcs/0049-flakes.md

```diff
@@ -0,0 +1,717 @@
+  ("flake references") directly in the `nix` command:
+
+  ```
+  # nix build github:NixOS/patchelf
+  ```
+
+* For a reproducible result, you can also use a specific revision:
+
+  ```
+  # nix build nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06#hello
+  ```
+
+* To get information about a flake:
+
+  ```
+  # nix flake info nixpkgs
+  Description:   A collection of packages for the Nix package manager
+  Revision:      a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ...
+  ```
+
+* In addition to the global registry, there is a per-user
+  registry. This can be used to pin flakes to the current version:
+  ```
+  # nix flake pin nixpkgs
+  ```
+  or to a specific version:
+  ```
+  # nix flake add nixpkgs github:edolstra/nixpkgs/a0e1f50e6f72e5037d71a0b65c67cf0605349a06
+  ```
+
+* Flakes can have dependencies on other flakes. For example, [the
+  `patchelf`
+  flake](https://github.com/NixOS/patchelf/blob/master/flake.nix)
+  depends on the `nixpkgs` flake. To ensure reproducibility,
+  dependencies are pinned to specific versions using a *lock file*
+  (e.g. [for
+  patchelf](https://github.com/NixOS/patchelf/blob/master/flake.lock)). Lock
+  files are generated automatically.
+
+* Flakes are evaluated in pure mode, meaning they can't access
+  anything other than their own source or declared dependencies. This
+  allows `nix` to cache evaluation results.
+
+* The `nix` command is flake-based. For example,
+  ```
+  # nix build
+  ```
+  builds the flake in the current directory.
+
+# Motivation
+[motivation]: #motivation
+
+Flakes are motivated by a number of serious shortcomings in Nix:
+
+* While Nix pioneered reproducible builds, sadly, Nix expressions are
+  not nearly as reproducible as Nix builds. Nix expressions can access
+  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
+  environment variables, and Git repositories. This means for instance
+  that it is hard to ensure reproducible evaluation of NixOS or NixOps
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
+
+* There is no standard way to compose Nix projects. Typical ways are
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems.
+
+# Detailed design
+[design]: #detailed-design
+
+## Flakes
+
+A flake is a Git repository that contains a file named `flake.nix` in
+the root directory. (In the future, there might be other types of
```

Comment:
Supporting tarballs seems like it's pretty necessary to me, especially since there's no pre-existing requirement to put nix files in a git repository (instead of e.g. mercurial, etc etc). On that point, it seems odd to merge in a possible RFC without even writing down the possibility of something non-git-based; especially since (to me) [it feels like](https://github.com/NixOS/rfcs/pull/49/files#r314726559) the only reason these are not in the spec is because it isn't yet implemented in nix.